Source/CarthageKit/Project.swift

```diff
@@ -1064,6 +997,21 @@ public final class Project { // swiftlint:disable:this type_body_length
 					return dependenciesIncludingNext
 				}
 			}
+			.flatMap(.concat) { (dependencies: [(Dependency, PinnedVersion)]) -> SignalProducer<[(Dependency, PinnedVersion)], CarthageError> in
+				return self.copyBuiltFrameworksFromSharedCacheIfVersionMatches(for: dependencies, options: options)
+					.flatMap(.concat) { dependency, version -> SignalProducer<(Dependency, PinnedVersion), CarthageError> in
+						self._projectEventsObserver.send(value: .skippedBuildingCopyFromSharedCache(dependency))
+						return SignalProducer(value: (dependency, version))
+					}
+					.collect()
+					.map { copiedFromSharedCacheDependencies -> [(Dependency, PinnedVersion)] in
+						// Filters out dependencies that we've copied from shared cache
+						// but preserves the build order
+						return dependencies.filter { dependency -> Bool in
+							!copiedFromSharedCacheDependencies.contains { $0 == dependency }
```

Comment:
This could be `!copiedFromSharedCacheDependencies.contains(dependency)`

Reply:
Impossible because tuple (Dependency, PinnedVersion) doesn't conforms to Equatable (same filter is done few lines below when downloading dependencies (Filters out dependencies that we've downloaded binaries for))